dpcpp/matrix/csr_kernels.dp.cpp

```diff
@@ -871,6 +871,72 @@ void extract_diagonal(size_type diag_size, size_type nnz,
 GKO_ENABLE_DEFAULT_HOST(extract_diagonal, extract_diagonal);
 
 
+template <typename IndexType>
+void check_diagonal_entries(const IndexType num_min_rows_cols,
+                            const IndexType* const __restrict__ row_ptrs,
+                            const IndexType* const __restrict__ col_idxs,
+                            bool* const __restrict__ has_all_diags,
+                            sycl::nd_item<3> item_ct1)
+{
+    constexpr int subgroup_size = config::warp_size;
+    auto tile_grp = group::tiled_partition<subgroup_size>(
+        group::this_thread_block(item_ct1));
+    const auto row =
+        thread::get_subwarp_id_flat<subgroup_size, IndexType>(item_ct1);
+    if (row < num_min_rows_cols) {
+        const auto tid_in_warp = tile_grp.thread_rank();
+        const auto row_start = row_ptrs[row];
+        const auto num_nz = row_ptrs[row + 1] - row_start;
+        bool row_has_diag_local{false};
+        for (IndexType iz = tid_in_warp; iz < num_nz; iz += subgroup_size) {
+            if (col_idxs[iz + row_start] == row) {
+                row_has_diag_local = true;
+                break;
+            }
+        }
+        auto row_has_diag = static_cast<bool>(tile_grp.any(row_has_diag_local));
+        if (!row_has_diag) {
+            if (tile_grp.thread_rank() == 0) {
+                *has_all_diags = false;
+            }
+            return;
+        }
+    }
+}
+
+GKO_ENABLE_DEFAULT_HOST(check_diagonal_entries, check_diagonal_entries);
+
+
+template <typename ValueType, typename IndexType>
+void add_scaled_identity(const ValueType* const __restrict__ alpha,
+                         const ValueType* const __restrict__ beta,
+                         const IndexType num_rows,
+                         const IndexType* const __restrict__ row_ptrs,
+                         const IndexType* const __restrict__ col_idxs,
+                         ValueType* const __restrict__ values,
+                         sycl::nd_item<3> item_ct1)
+{
+    constexpr int subgroup_size = config::warp_size;
+    auto tile_grp = group::tiled_partition<subgroup_size>(
+        group::this_thread_block(item_ct1));
+    const auto row =
+        thread::get_subwarp_id_flat<subgroup_size, IndexType>(item_ct1);
+    if (row < num_rows) {
+        const auto tid_in_warp = tile_grp.thread_rank();
+        const auto row_start = row_ptrs[row];
+        const auto num_nz = row_ptrs[row + 1] - row_start;
+        for (IndexType iz = tid_in_warp; iz < num_nz; iz += subgroup_size) {
+            values[iz + row_start] *= beta[0];
+            if (col_idxs[iz + row_start] == row) {
+                values[iz + row_start] += alpha[0];
+            }
+        }
```

Comment:
It might make sense to check the scaling value itself first before performing these operations (in case of beta == 1, you don't have to read and write most of the values).
```suggestion
        const auto num_nz = row_ptrs[row + 1] - row_start;
        const auto beta_v = beta[0];
        const auto alpha_v = alpha[0];
        for (IndexType iz = tid_in_warp; iz < num_nz; iz += subgroup_size) {
            if (beta_v != one<ValueType>()) {
                values[iz + row_start] *= beta_v;
            }
            if (col_idxs[iz + row_start] == row) {
                if (alpha_v != zero<ValueType>()) {
                    values[iz + row_start] += alpha_v;
                }
            }
        }
```

Reply:
I didn't check it, but the condition should be the same for all threads, so I was hoping it would be optimized properly.
In any case, if you only want to scale the diagonal, the check should significantly improve performance (maybe the `alpha` check is not that necessary, but the `beta` check does make sense IMO).

Reply:
agree. I think alpha check does not hurt because we already have the check for diagonal